src/de.rs

```diff
@@ -1,29 +1,45 @@
 use pyo3::types::*;
+use pyo3::{Py, PyObject};
 use serde::de::{self, IntoDeserializer};
 use serde::Deserialize;
 use std::convert::TryInto;
+use std::marker::PhantomData;
 
 use crate::error::{PythonizeError, Result};
+use crate::types::*;
+use crate::{DefaultPythonizeTypes, PythonizeTypes};
 
 /// Attempt to convert a Python object to an instance of `T`
-pub fn depythonize<'de, T>(obj: &'de PyAny) -> Result<T>
+pub fn depythonize<'py, T>(obj: &'py PyAny) -> Result<T>
 where
-    T: Deserialize<'de>,
+    T: Deserialize<'py>,
 {
-    let mut depythonizer = Depythonizer::from_object(obj);
-    T::deserialize(&mut depythonizer)
+    depythonize_custom::<DefaultPythonizeTypes, _>(obj)
 }
 
-pub struct Depythonizer<'de> {
-    input: &'de PyAny,
+/// Attempt to convert a Python object to an instance of `T` using custom python types
+pub fn depythonize_custom<'py, P, T>(obj: &'py PyAny) -> Result<T>
+where
+    T: Deserialize<'py>,
+    P: PythonizeTypes<'py> + 'py,
+{
+    T::deserialize(Depythonizer::<P>::from_object(obj))
+}
+
+pub struct Depythonizer<'py, P: PythonizeTypes<'py>> {
+    input: &'py PyAny,
+    _pythonize: PhantomData<*mut &'py P>,
```

Comment:
It is needed to be like that, because of the covariance rules:
https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns

It is relevant, because T here might be a reference

Reply:
I don't follow - if `P` is `&DefaultPythonizeTypes`, for example, then `PhantomData<P>` already expands to `PhantomData<&DefaultPythonizeTypes>` and the right rules therefore still apply?

Can you provide a counterexample where `PhantomData<P>` does the wrong thing here?

Reply:
No they dont. We actually dont containt reference to the `DefaultPythonizeType`